cpp/src/streaming/core/memory_reserve_or_wait.cpp

```diff
@@ -0,0 +1,229 @@
+        -> std::ranges::subrange<std::set<ResReq>::const_iterator> {
+        // Since `reservation_requests_` is sorted by ascending size,
+        // upper_bound finds the first element with size > max_size.
+        auto last = std::ranges::upper_bound(
+            reservation_requests_, max_size, std::less<>{}, &ResReq::size
+        );
+        // The range [begin, last) contains all requests with size <= max_size.
+        return {reservation_requests_.begin(), last};
+    };
+
+    // Helper that pushes a memory reservation into a request's queue **without**
+    // waiting on the coroutine.
+    auto push_into_queue = [this](ResReq& request, MemoryReservation res) -> void {
+        auto err = ctx_->executor()->spawn_detached(
+            [](coro::queue<MemoryReservation>& queue, MemoryReservation res) -> Node {
+                RAPIDSMPF_EXPECTS(
+                    co_await queue.push(std::move(res))
+                        == coro::queue_produce_result::produced,
+                    "could not push memory reservation"
+                );
+            }(request.queue, std::move(res))
+        );
+        RAPIDSMPF_EXPECTS(err, "cannot spawn push-into-queue task");
+    };
+
+    while (true) {
+        auto last_reservation_success = Clock::now();
+        while (true) {
+            // Exit if no more pending requests remain.
+            {
+                std::unique_lock lock(mutex_);
+                if (reservation_requests_.empty()) {
+                    co_return;
+                }
+            }
+            periodic_memory_check_counter_.fetch_add(1, std::memory_order_acq_rel);
+            co_await ctx_->executor()->yield();
+            if (Clock::now() - last_reservation_success > timeout_) {
+                // This is the only way out of the while-loop that doesn't shutdown
+                // the periodic memory check.
+                break;
+            }
+            auto const max_size = memory_available();
+
+            // Find the request with the greatest future_release_potential that fits
+            // into the currently available memory.
+            std::unique_lock lock(mutex_);
+            auto eligibles = eligible_requests(max_size);
+            if (eligibles.begin() == eligibles.end()) {
+                continue;  // No eligible requests.
+            }
+            auto it = std::ranges::max_element(
+                eligibles, std::less<>{}, &ResReq::future_release_potential
+            );
+
+            // Try to reserve memory for the selected request.
+            auto [res, _] = ctx_->br()->reserve(mem_type_, it->size, no_overbooking);
+            if (res.size() == 0) {
+                continue;  // Memory is no longer available.
+            }
+
+            // Extract the selected request and push the reservation into its queue.
+            ResReq request = reservation_requests_.extract(it).value();
+            lock.unlock();
+            push_into_queue(request, std::move(res));
```

Comment:
*issue*: Maybe? `push_into_queue` takes the request by reference. But that reference could be dead by the time the queue is accessed inside the task?